### ISSUE TITLE
Clarify Collector's Describe(): Returning all possible metrics required? When not?

### ISSUE THREAD

**Issue by discordianfish** [prometheus/client_golang#140]
The [Collector interface](https://godoc.org/github.com/prometheus/client_golang/prometheus#Collector) describes `Describe()` as:

```
Describe sends the super-set of all possible descriptors of metrics
collected by this Collector to the provided channel and returns once
the last descriptor has been sent.
```

In reality this often is hard to achieve, therefor we have lot of exporters (e.g graphite exporter) simply not doing that but returning just a subset of possible metric descriptors.

Even to me it's still unclear why this is necessary and how does it help with consistency given that it doesn't enforce the described requirements.

**Comment by beorn7** [comment]
tl;dr: This is essentially a duplicate of https://github.com/prometheus/client_golang/issues/47

If you call `prometheus.EnableCollectChecks(true)`, it _will_ enforce the requirements.
If you do not call it, the requirement is just a contract. Bug-free collectors must honor it, so we can save the scrape-time overhead of checking it.

Also, the whole reason for having `Describe()` is to check consistency upon registration time (i.e. when you start up your program) and not upon scrape time. Panicking upon startup is kind of acceptable while panicking upon touching an HTTP endpoint is not.

**Comment by discordianfish** [comment]
OH forgot about that. Thanks. I opened this because the author of heka-prometheus was confused about that and had crazy things in mind to workaround that.
